runners/flink/src/main/java/org/apache/beam/runners/flink/translation/wrappers/streaming/DoFnOperator.java

```diff
@@ -837,11 +837,15 @@ public void onProcessingTime(InternalTimer<ByteBuffer, TimerData> timer) {
 
   // allow overriding this in ExecutableStageDoFnOperator to set the key context
   protected void fireTimerInternal(Object key, TimerData timerData) {
-    fireTimer(timerData);
+    if (key instanceof java.nio.ByteBuffer) {
+      key = FlinkKeyUtils.decodeKey((ByteBuffer) key, keyCoder);
+    }
+
+    fireTimer(key, timerData);
   }
 
   // allow overriding this in WindowDoFnOperator
-  protected void fireTimer(TimerData timerData) {
+  protected void fireTimer(Object key, TimerData timerData) {
```

Comment:
```suggestion
  protected void fireTimer(ByteBuffer key, TimerData timerData) {
```

Reply:
If you decide the decode the key already in the calling method, then use the key type here.